### PR TITLE
TINKERPOP-1312 .count().is(0) is not properly optimized

### PR DESCRIPTION
**Pull request by dkuppitz**
Tweaked `RangeByIsCountStrategy`. The patterns ...

* `....count().is(0)`
* `....count().is(lt(1))`
* `....count().is(lte(0))`

... are now replaced by `not(...)`.

`docker/build.sh -t -i -n` succeeded.

VOTE: +1

### CODE REVIEW

**Comment by okram** [comment]
Can you add some more complex tests? For instance:

```
out().count().is(0).as("a")
```

Also, `RangeByIsCountStrategy` doesn't have `applyPrior()` of `AdjacentToIncidentStrategy`. Does that matter? That is:

```
out().count().is(0) ->
outE().count().is(0) ->
not(outE())
```

We want `not(outE())` NOT `not(out())`. Can you verify that such compilations happen?

Also, why do you make a distinction between computer and standard engines for this strategy?

I think you should copy that pattern in the other strategy tests that ensure proper `__()` compilations. We can go crazy on adding patterns to make sure compilations are as expected. E.g.:

```
  @Parameterized.Parameters(name = "{0}")
    public static Iterable<Object[]> generateTestParameters() {

        return Arrays.asList(new Traversal[][]{
                {__.outE().count(), __.outE().count()},
                {__.bothE("knows").count(), __.bothE("knows").count()},
                {__.properties().count(), __.properties().count()},
                {__.properties("name").count(), __.properties("name").count()},
                {__.out().count(), __.outE().count()},
                {__.in().count(), __.inE().count()},
                {__.both().count(), __.bothE().count()},
                {__.out("knows").count(), __.outE("knows").count()},
                {__.out("knows", "likes").count(), __.outE("knows", "likes").count()},
                {__.filter(__.out()), __.filter(__.outE())},
                {__.where(__.not(__.out())), __.where(__.not(__.outE()))},
                {__.where(__.out("knows")), __.where(__.outE("knows"))},
                {__.values().count(), __.properties().count()},
                {__.values("name").count(), __.properties("name").count()},
                {__.where(__.values()), __.where(__.properties())},
                {__.and(__.out(), __.in()), __.and(__.outE(), __.inE())},
                {__.or(__.out(), __.in()), __.or(__.outE(), __.inE())},
                {__.out().as("a").count(), __.outE().count()},   // TODO: is this good?
                {__.where(__.as("a").out("knows").as("b")), __.where(__.as("a").out("knows").as("b"))}});
    }
```

**Comment by dkuppitz** [comment]
> why do you make a distinction between computer and standard engines for this strategy's test case?

I was wondering why we initially chose another pattern for this test class; I couldn't figure it out, but then still just followed the pattern that we started with. I'll change that.

**Comment by dkuppitz** [comment]
Updated. Regarding the order of `RangeByIsCountStrategy` and `AdjacentToIncidentStrategy`, it doesn't really matter. We either go:

`out().count().is(0) -> outE().count().is(0) -> not(outE())`

or:

`out().count().is(0) -> not(out()) -> not(outE())`

In fact it's the former, but either way we would always end up with the same step sequence.

**Comment by okram** [comment]
Sorry to be a pest, but could you add more complex tests still! :)

```
outE().count().is(0).inV()
both().both().count().is(0).as("a")
```

For the latter, should it be `not(both().both())` ? What are the semantics in longer chains -- also, what about local children and global children. All your test cases are basically "the same" in terms of just `out().*`.

**Comment by dkuppitz** [comment]
> For the latter, should it be `not(both().both())`?

Since we only look at it in conjunction with `count().is(something)`, we don't need to differentiate between single steps and longer chains.

`both().not(both())` is the same thing as `not(both().both())`.

It makes a difference if something comes after `is(0)`, but in that case we don't use `not()`:

`both().both().count().is(0).both() -> both().both().limit(1).count().is(0).both()`

**Comment by okram** [comment]
When all is in place, can you provide a benchmark on Grateful Dead graph of some query in master/ vs. this branch and their runtimes? So to be certain its all as expected. Even provide a `explain()` of the traversal you are benchmarking so we see how it compiled?

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/RangeByIsCountStrategyTest.java [hunk 1]
```diff
@@ -142,6 +142,7 @@ public void doTest(final Traversal traversal, final Traversal optimized) {
                     {__.out().count().is(without(2, 6, 4)), __.out().limit(6).count().is(without(2, 6, 4))},
                     {__.map(__.count().is(0)), __.map(__.limit(1).count().is(0))},
                     {__.flatMap(__.count().is(0)), __.flatMap(__.limit(1).count().is(0))},
+                    {__.filter(__.count().is(0)), __.filter(__.not(__.identity()))},
```
Comment (okram) [inline gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/RangeByIsCountStrategyTest.java]: `filter(not(_.identity))` the same as `not()` ?

**Comment by okram** [comment]
Sweet benchmark. VOTE +1.

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/RangeByIsCountStrategyTest.java (hunk 1 above)
Comment (dkuppitz) [inline gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/RangeByIsCountStrategyTest.java]: It is the same, but IMO it's not RangeByIsCountStrategy's job to optimize that.

**Comment by spmallette** [comment]
Nice tests and benchmark stuff  - VOTE +1

**Comment by dkuppitz** [comment]
Not sure why the merge was not recognized automatically, but it's merged into `tp31` and `master`.

**Pull request closed without merging**